File: src/main/java/no/mesan/fag/patterns/timesheet/external/decorators/TimeDataServiceLoggingDecorator.java
package no.mesan.fag.patterns.timesheet.external.decorators;

import java.util.List;
import java.util.logging.Logger;

import no.mesan.fag.patterns.timesheet.data.TimesheetEntry;
import no.mesan.fag.patterns.timesheet.external.TimeDataService;

/// HINT starten på en klasse som kan logge kall til tjenesten
/** Dekoratør for den eksterne tjenesten som legger på litt fancy logging. */
public class TimeDataServiceLoggingDecorator extends TimeDataServiceDecorator {

    private Logger logger = Logger.getLogger("TimeDataServiceLoggingDecorator");

    public TimeDataServiceLoggingDecorator(final TimeDataService cachingDecoratedService) {
        super(cachingDecoratedService);
    }


    /// HINT type logging vi ønsker
    // Før kall: logger.info(String.format("Parameters: userID=%s, from=%d", userID, from));
    // Etter kall:  logger.info(String.format("Found %d entries, in %dms", timesheetEntries.size(), timeSpent));
    @Override
    public List<TimesheetEntry> forEmployee(final String userID, final int from) {
        return null;
    }

    @Override
    public List<TimesheetEntry> forYear(final int year, final int from) {
        return null;
    }

    /** For test. */
    void setLogger(final Logger logger) {
        this.logger = logger;
    }
}

File: src/main/scala/no/mesan/fag/patterns/scala/timesheet/external/decorators/TimeDataServiceLoggingDecorator.scala
package no.mesan.fag.patterns.scala.timesheet.external.decorators

import no.mesan.fag.patterns.scala.timesheet.external.TimeDataService
import no.mesan.fag.patterns.scala.timesheet.data.TimesheetEntry

import java.util.logging.Logger

/** Dekoratør for den eksterne tjenesten som legger på litt fancy logging. */
trait TimeDataServiceLoggingDecorator extends TimeDataService {
  // Loggeren
  private var logger = Logger.getLogger("TimeDataServiceLoggingDecorator")

  protected[decorators] def setLogger(newLogger:Logger) { logger= newLogger }

  private def runWithTimer(intro:String)(f: => List[TimesheetEntry]): List[TimesheetEntry] = {
    val startsAt = System.nanoTime
    val result= f
    val endsAt= System.nanoTime
    logger.info(s"$intro\nFound ${result.size} entries in ${(endsAt-startsAt)/1e6}ms")
    result
  }

  /// HINT overstyr interfacemetodene ved å kalle dem via foregående
}
